### PR TITLE
Add waveguide library and mode utilities

### PR DESCRIPTION
**Pull request by lucas-flexcompute**

### CODE REVIEW

**Comment by lucas-flexcompute** [comment]
[Here's a sample notebook](https://gist.github.com/lucas-flexcompute/488b3ecb2993a67417a22f22900b0866) using the added functionality.

**Review by tylerflex — COMMENTED** [review]
@lucas-flexcompute this seems like it would be really useful! I just have a few questions / comments:

the generalized dot products should certainly be useful, but how do you imagine the `rectangular_dielectric` interfacing with the rest of the tidy3D workflow?  Right now it seems well set up for separate use as it's basically a wrapper around `td.plugins.mode.ModeSolver`. What about the situation where a user has generated a `ModeSolver` using `rectangular_dielectric` and now they'd like to export some components (sources, structures, or ModeSpec, for example) for use in a Tidy3D simulation? Do you anticipate it being integrated like this or used completely separately?

File: tidy3d/components/data/monitor_data.py [hunk 1]
```diff
@@ -374,6 +420,103 @@ def dot(
...
+                    h_self_x_e_other = h_self_1 * e_other_2 - h_self_2 * e_other_1
+
+                    # Integrate over plane
+                    d_area = self._diff_area
+                    integrand = (e_self_x_h_other - h_self_x_e_other) * d_area
+                    dot[i, mi0, mi1] = 0.25 * integrand.sum(dim=d_area.dims)
+
+        coords = {"f": f, "mode_index_0": mode_index_0, "mode_index_1": mode_index_1}
+        return MixedModeDataArray(dot, coords=coords)
+
```
Comment (momchil-flex) [inline tidy3d/components/data/monitor_data.py:519]: This method seems unnecessarily long/complicated compared to how I would have gone about it, and also there's a lot of repeated code with the regular `dot`, which we always try to avoid. After thinking about it for a little bit, it seems like the main drawback is that we don't have `FieldData` methods to `interp` things and return another `FieldData` - instead you have to manually re-collect xarray DataArrays into a `FieldData`. We've gone back and forth on this a few times, whether we want to wrap these methods, and last time we decided not to, one reason being that we don't want `FieldData.monitor` to become stale w.r.t. the data stored in the data structure.

Still, it seems to me that a better approach would be something like:
- Get centered tangential fields of `self`
- Create a new `FieldData` with a monitor given by `other.monitor.copy(update=dict(colocate=True))`, and with fields given by the `other` fields `interp`-ed to the centers of `self`
- Iterate over the common frequencies and `i = range(self.mode_index.size)`, and do something likel `self._isel(mode_index=i).dot(colocated_other)`.

`_isel` is a method we do have implemented already for internal use. Maybe we could implement `_interp` too if it will simplify things.

**Review by momchil-flex — COMMENTED** [review]
I think I agree with Tyler, we need to clarify how to structure the `rectangular_dielctric` scope and interplay with other tools. For example, as per Tyler's question, would it ever be useful in an FDTD simulation, not just in the mode solver? If yes, then it may be better to have it as a more standalone feature (a class probably) that can be used to initialize both a mode solver and a simulation.

File: tidy3d/components/data/monitor_data.py [hunk 2]
```diff
@@ -320,6 +350,22 @@ def flux(self) -> FluxDataArray:
         d_area = self._diff_area
         return FluxDataArray((self.poynting * d_area).sum(dim=d_area.dims))
 
+    @cached_property
+    def mode_area(self) -> FreqModeDataArray:
+        """Effective mode area corresponding to a 2D monitor
+
+        Effective mode area is calculated as: (∫|E|²dA)² / (∫|E|⁴dA)
+        """
+        fields = self._centered_fields
```
[the remark this answers is not in the document]
Comment (momchil-flex) [reply tidy3d/components/data/monitor_data.py:359]: Sorry, I see why. I think in the interest of avoiding repeating code, we should reorganize things though. We can have `_tangential_centered_fields` take `_centered_fields` and pick out the tangential components.

[the remark this answers is not in the document]
Comment (lucas-flexcompute) [reply tidy3d/components/data/monitor_data.py:359]: Because it also needs the longitudinal components.

File: tidy3d/components/data/monitor_data.py (hunk 1 above)
Comment (lucas-flexcompute) [reply tidy3d/components/data/monitor_data.py]: When other is a `ModeSolverMonitor`, `colocate` is defined to be always false, so I guess `other.monitor.copy(update=dict(colocate=True))` won't work.

I've refactored much of this code in the current commit, but not in this manner you mention yet.

**Comment by lucas-flexcompute** [comment]
Note: [the sample notebook](https://gist.github.com/lucas-flexcompute/488b3ecb2993a67417a22f22900b0866) is also updated

File: tidy3d/components/data/monitor_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (momchil-flex) [reply tidy3d/components/data/monitor_data.py:371]: Yes re sign inversion; re grid correction, you should apply the `primal` grid correction to the normal H field, and the `dual` grid correction to the normal E field because that's where these components live along that axis.

File: tidy3d/components/data/monitor_data.py (hunk 1 above)
Comment (momchil-flex) [reply tidy3d/components/data/monitor_data.py]: Yeah I feel like this whole handling maybe needs a rethinking, but also it feels like a low priority compared to other things we're working on now... The usual dilemma. I'll have a look when you think you have a final (for now) version though.

File: tidy3d/components/data/monitor_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucas-flexcompute) [reply tidy3d/components/data/monitor_data.py:371]: Thanks, I'll fix the corrections. This sign inversion for the purpose of inverting the propagation direction direction might be a problem in the future, though. When fully anisotropic magnetic materials are implemented, it is possible to build unidirectional waveguide modes, so the sign cannot be not freely swapped. I guess this is a problem for later, but we should remember it.

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+    - Strip and rib geometries
+    - Angled sidewalls
+    - Modes in waveguide bends
+    - Surface and sidewall loss models
+    - Coupled waveguides
+    """
+
+    wavelength: Union[float, ArrayLike[float, 1]] = pydantic.Field(
+        ...,
+        description="Wavelength(s) at which to calculate modes (in μm).",
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:36]: add `title` fields as well (mainly for rendering in docs and GUI)

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+        ...,
+        description="Medium associated with the upper cladding layer.",
+    )
+
+    box_medium: MediumType = pydantic.Field(
+        None,
+        description="Medium associated with the lower cladding layer.",
+    )
+
+    slab_thickness: float = pydantic.Field(
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:68]: probably `pydantic.NonNegativeFloat` and wherever else this might apply.

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+    def _set_box_thickness(cls, val, values):
+        if val is None:
+            wavelength = values["wavelength"]
+            medium = values["box_medium"]
+            n = numpy.array([medium.nk_model(f)[0] for f in C_0 / wavelength])
+            lda = wavelength / n
+            return 1.5 * lda.max()
+        return val
+
+    @pydantic.validator("side_margin", always=True)
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:199]: might be simpler and clearer to let `side_margin` (and other similar things) be `@cached_property`s

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+            def polyslab_vertices(x, w):
+                return (
+                    self._transform_in_plane(x, -half_length),
+                    self._transform_in_plane(x + w, -half_length),
+                    self._transform_in_plane(x + w, half_length),
+                    self._transform_in_plane(x, half_length),
+                )
+
+        else:
+            assert (
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:373]: better to raise rather than assert

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+    @cached_property
+    def width(self):
+        """Domain width (size in the lateral direction)"""
+        w = self.core_width.sum() + self.gap.sum() + 2 * self.side_margin
+        if self.sidewall_angle > 0:
+            w += 2 * self.core_thickness * numpy.tan(self.sidewall_angle)
+        return w
+
+    # pylint:disable=too-many-locals,too-many-statements,too-many-branches
+    @cached_property
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:267]: this method should be broken up I think.

**Review by tylerflex — CHANGES_REQUESTED** [review]

File: tidy3d/plugins/waveguide/rectangular_dielectric.py (hunk 5 above)
Comment (lucas-flexcompute) [reply tidy3d/plugins/waveguide/rectangular_dielectric.py:199]: Is this what you mean?
```python
    @cached_property
    def clad_thickness(self):
        if "clad_thickness" in self.__fields_set__:
            return self.__fields_set__["clad_thickness"]
        wavelength = self.wavelength
        medium = self.clad_medium
        n = numpy.array([medium.nk_model(f)[0] for f in C_0 / wavelength])
        lda = wavelength / n
        return 1.5 * lda.max()
```

File: tidy3d/plugins/waveguide/rectangular_dielectric.py (hunk 7 above)
Comment (lucas-flexcompute) [reply tidy3d/plugins/waveguide/rectangular_dielectric.py:267]: This function is long because it creates the structures and override structures for the mode solver for any setting the user might pass, so it handles any cases the class supports. The thing is that each part of the code depends on the variables calculated before it, so, if we break it up into smaller functions, we'd have to bass a bunch of variables around and that list would just keep increasing along the calculation. Furthermore, I don't think there's any piece of code here that could be reused somewhere else. Breaking this up would mean calling a series of different functions in a precise order passing a lot of arguments that would lose their meaning in the context of an isolated function, so the result would be an increase in complexity, in my opinion.

**Comment by lucas-flexcompute** [comment]
I've added an example notebook to our docs regarding this plugin: https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/lucas/waveguides/docs/source/notebooks/WaveguidePlugin.ipynb

I'm starting to wonder whether we should wait for a solution for the sub-pixel smoothing in the mode solver before merging this because as it stands the mode solver is not accurate enough for calculating the group index of a waveguide nor bend losses (see input cell 17 in the example notebook), for example.

I'll probably continue to work on it when time permits to include optical fiber profiles and factoring out a base class for general waveguide cross-sections, anyway.

File: tidy3d/components/data/monitor_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (momchil-flex) [reply tidy3d/components/data/monitor_data.py:380]: The sign inversion here is not for the purpose of propagation direction inversion, but rather because we are working in (x, z, y) reference frame, which is left-handed. Your general comment is definitely valid for the case of computing expansion coefficients though, as there we do not re-compute the backward propagating ones currently, but just use the time-reversed version of the forward propagating ones. Also, this conversion to and from a left-handed frame when the normal axis is y may need something extra in that case, namely some of the off-diagonal eps components will have to get a - sign?

I guess it would be great to improve this handling in general... We have a number of places in the code where we check `if normal_ind == 1`. I kinda regret not working in (z, x, y) reference frame from the start, but changing to that now is its own can of worms.

**Comment by momchil-flex** [comment]
> I've added an example notebook to our docs regarding this plugin: https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/lucas/waveguides/docs/source/notebooks/WaveguidePlugin.ipynb
> 
> I'm starting to wonder whether we should wait for a solution for the sub-pixel smoothing in the mode solver before merging this because as it stands the mode solver is not accurate enough for calculating the group index of a waveguide nor bend losses (see input cell 17 in the example notebook), for example.
> 
> I'll probably continue to work on it when time permits to include optical fiber profiles and factoring out a base class for general waveguide cross-sections, anyway.

Sounds good. Would you expect the loss to go monotonically down with bend radius?

**Comment by lucas-flexcompute** [comment]
> > I've added an example notebook to our docs regarding this plugin: https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/lucas/waveguides/docs/source/notebooks/WaveguidePlugin.ipynb
> > I'm starting to wonder whether we should wait for a solution for the sub-pixel smoothing in the mode solver before merging this because as it stands the mode solver is not accurate enough for calculating the group index of a waveguide nor bend losses (see input cell 17 in the example notebook), for example.
> > I'll probably continue to work on it when time permits to include optical fiber profiles and factoring out a base class for general waveguide cross-sections, anyway.
> 
> Sounds good. Would you expect the loss to go monotonically down with bend radius?

Yes. I just tried increasing the precision to double and the grid density ant it actually behaved as expected. I'll run e few more tests, but maybe the grid+double precision are enough for now.

**Comment by lucas-flexcompute** [comment]
I've revised the defaults for the waveguide plugin and, for now, the precision is adequate for the computations we're doing. When we have sub-pixel smoothing in the mode solver, we might be able to reduce the resolution and keep the accuracy. The [accompanying notebook](https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/lucas/waveguides/docs/source/notebooks/WaveguidePlugin.ipynb) has been updated with these new defaults.

File: tidy3d/components/data/monitor_data.py [hunk 8]
```diff
@@ -374,6 +441,114 @@ def dot(
...
+                    # Cross products of fields
+                    e_self_x_h_other = e_self_1 * h_other_2 - e_self_2 * h_other_1
+                    h_self_x_e_other = h_self_1 * e_other_2 - h_self_2 * e_other_1
+
+                    # Integrate over plane
+                    d_area = self._diff_area
+                    integrand = (e_self_x_h_other - h_self_x_e_other) * d_area
+                    dot[i, mi0, mi1] = 0.25 * integrand.sum(dim=d_area.dims)
+
+        coords = {"f": f, "mode_index_0": mode_index_0, "mode_index_1": mode_index_1}
```
Comment (momchil-flex) [inline tidy3d/components/data/monitor_data.py:549]: I guess this is fine, but there's still quite some overlapping code between this method and `dot`. Did you have any particular reason against going the route of using (custom) `isel` type methods and direct calls to `dot`?

**Review by momchil-flex — COMMENTED** [review]
I had a look at the data modifications, I'm leaving the review of the actual plugin to @tylerflex .

Something important given the data modifications: you should run the backend tests either on your computer or using the github action, in order to make sure that nothing has been broken. Let me know if you need help with that.

**Comment by momchilmm** [comment]
Actually one more note: it seems to me that your `centered_fields` handling does not check if the fields are already colocated (i.e. `FieldData.monitor.colocate`)?

File: tidy3d/components/data/monitor_data.py (hunk 8 above)
Comment (lucas-flexcompute) [reply tidy3d/components/data/monitor_data.py:549]: There was the issue with setting `colocate = True` for `ModeSolverMonitor`, which doesn't work and, tbh, creating a separate monitor and doing custom selections to get `dot` to perform the operation between the desired modes seemed like increasing complexity to me, compared to just doing the operation in place. I would actually consider the other way around, having `dot` as a special case of `outer_dot`, but then we also lose performance doing unnecessary interpolations. In the end (besides the monitor issue) it just felt like the code duplication here is minimal compared to complexity cost, but I can change that if you don't want the duplication here. If that's the case, can I just make a `ModeMonitor` from a `ModeSolverMonitor`?

Comment (momchil-flex) [reply tidy3d/components/data/monitor_data.py:548]: Ok, let's leave it like this for now, maybe I'll rethink it later when making the changes to colocate by default.

File: tidy3d/plugins/waveguide/rectangular_dielectric.py (hunk 5 above)
Comment (tylerflex) [reply tidy3d/plugins/waveguide/rectangular_dielectric.py:233]: do we intend to set clad_thickness directly in the plugin? if so ignore what I said about property. What I was getting at is that if the clad thickness is just a derived property, we should do it as a property instead of as a validator

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 9]
```diff
@@ -0,0 +1,944 @@
...
+    def _validate_gaps(cls, val, values):
+        if val.size == 1 and values["core_width"].size != 2:
+            return numpy.array([val[0]] * (values["core_width"].size - 1))
+        if val.size != values["core_width"].size - 1:
+            raise ValidationError("Number of gaps must be 1 less than number of core widths.")
+        return val
+
+    @pydantic.root_validator
+    def _ensure_consistency(cls, values):
+        if values["sidewall_thickness"] > 0 and values["sidewall_medium"] is None:
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:247]: maybe for readability, just assign `values[...]` to variables at the beginning of this.

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 10]
```diff
@@ -0,0 +1,944 @@
...
+            lda = wavelength / n
+            return 1.5 * lda.max()
+        return val
+
+    @pydantic.validator("side_margin", always=True)
+    def _set_side_thickness(cls, val, values):
+        return max(values["clad_thickness"], values["box_thickness"]) if val is None else val
+
+    @pydantic.validator("gap", always=True)
+    def _validate_gaps(cls, val, values):
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:238]: can we add docstrings to the validators describing what they are checking?

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 11]
```diff
@@ -0,0 +1,944 @@
...
+            log.warning("Surface medium not used because surface thickness is zero.")
+
+        if values["propagation_axis"] == values["normal_axis"]:
+            raise ValidationError("Propagation and normal axes must be different.")
+
+        return values
+
+    @cached_property
+    def lateral_axis(self):
+        """Lateral direction axis"""
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:270]: period at the end of docstring

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 12]
```diff
@@ -0,0 +1,944 @@
...
+        if values["surface_thickness"] == 0 and values["surface_medium"] is not None:
+            log.warning("Surface medium not used because surface thickness is zero.")
+
+        if values["propagation_axis"] == values["normal_axis"]:
+            raise ValidationError("Propagation and normal axes must be different.")
+
+        return values
+
+    @cached_property
+    def lateral_axis(self):
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:269]: add return type

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 13]
```diff
@@ -0,0 +1,944 @@
...
+        """Lateral direction axis"""
+        return 3 - self.propagation_axis - self.normal_axis
+
+    def _swizzle(self, lateral_coord, normal_coord, propagation_coord):
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:273]: interesting method name 😆 . I like it, but would maybe suggest that we use something more boring but also more descriptive. `_set_waveguide_coords`? Also, add type hints please to this and the other ones. and add a periodic at the end of docstrings.

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 14]
```diff
@@ -0,0 +1,944 @@
...
+        result[self.normal_axis] = normal_coord
+        return result
+
+    def _translate(self, lateral_coord, normal_coord, propagation_coord):
+        """Swap the model coordinates to desired axes and translate to origin"""
+        result = [
+            a + b
+            for a, b in zip(
+                self.origin,
+                self._swizzle(lateral_coord, normal_coord, propagation_coord),
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:287]: for readability, let's assign `self._swizzle(lateral_coord, normal_coord, propagation_coord)` to a variable like `model_coordinate`?

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 15]
```diff
@@ -0,0 +1,944 @@
...
+            )
+        ]
+        return result
+
+    def _transform_in_plane(self, lateral_coord, propagation_coord):
+        """Swap the model coordinates to desired axes in the substrate plane"""
+        result = self._translate(lateral_coord, 0, propagation_coord)
+        result.pop(self.normal_axis)
+        return result
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:296]: you can use
```py
_, result = Box.pop_axis(result, axis=self.normal_axis)
```
to be consistent, in case it changes later for some reason.

File: tidy3d/plugins/waveguide/rectangular_dielectric.py (hunk 7 above)
Comment (tylerflex) [reply tidy3d/plugins/waveguide/rectangular_dielectric.py:312]: I'm not sure I agree actually. It seems to me that the function has a series of pretty logically distinct operations, which could be captured into their own methods. At the very least, probably setting up the grid spec, and creating the structures is the first step (the method name even mentions doing two things). But I think it would help to be refactored with more granularity.  My worry is that in its current state, this method would be a bit difficult for another developer to come in and extend / edit. It also doesn't lend itself well to testing each step of the process. Do you have a good feeling for how much complexity refactoring this method would introduce? To avoid passing too many arguments, I think adding some functions defined locally to `_structures_and_gridspec()` could be an option.

Comment (tylerflex) [reply tidy3d/plugins/waveguide/rectangular_dielectric.py:312]: Also, what are your ideas for the future of this plugin? do you plan to extend it in any way? or use it in other plugins?

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 16]
```diff
@@ -0,0 +1,944 @@
...
+            min_steps_per_wvl=self.grid_resolution,
+            wavelength=self.wavelength.min(),
+            override_structures=override_structures,
+            max_scale=self.max_grid_scaling,
+        )
+
+        if self.mode_spec.bend_radius is None or self.mode_spec.bend_radius == 0.0:
+            half_length = 0.5 * self.length
+
+            def polyslab_vertices(x, w):
```
Comment (tylerflex) [inline tidy3d/plugins/waveguide/rectangular_dielectric.py:442]: add docstrings and type hint (and below)

**Review by tylerflex — CHANGES_REQUESTED** [review]
Hey Lucas, thanks for this plugin. Overall I think it looks really good, I had a few comments and we can discuss the refactoring of the long method. My main concern is that some bits of the code have a lot going on and are a bit hard to follow as a result. However, I also understand that this is also just a messy problem to tackle, so we should accept some of complexity in the code. I tried to write a comment wherever I got a bit confused so we can discuss in the review comments, but the overall idea is good and excited to include this feature. Thanks!

File: tidy3d/components/data/monitor_data.py [hunk 17]
```diff
@@ -194,34 +200,26 @@ def _tangential_dims(self) -> List[str]:
...
-        return tan_fields
+    def _grid_corrected_fields(self) -> Dict[str, DataArray]:
+        """Dictionary of field components with finite grid correction factors applied and symmetry
+        expanded."""
+        if len(self.monitor.zero_dims) != 1:
+            raise DataError("Data must be 2D to apply grid corrections.")
+        normal_dim = "xyz"[self.monitor.zero_dims[0]]
+        fields = {}
+        for comp, field in self.symmetry_expanded_copy.field_components.items():
+            if comp[1] == normal_dim:
```
[the remark this answers is not in the document]
Comment (lucas-flexcompute) [reply tidy3d/components/data/monitor_data.py:211]: Nice, I didn't know about `symmetry_eigenvalues`!

File: tidy3d/components/data/monitor_data.py [hunk 18]
```diff
@@ -245,11 +242,17 @@ def _plane_grid_boundaries(self) -> Tuple[Coords1D, Coords1D]:
...
+            fields = self.symmetry_expanded_copy.field_components
+            plane_bounds1 = fields["E" + dim2].coords[dim1].values
+            plane_bounds2 = fields["E" + dim1].coords[dim2].values
 
         return plane_bounds1, plane_bounds2
 
+    @property
+    def _plane_grid_centers(self) -> Tuple[Coords1D, Coords1D]:
+        """For 2D monitor data, return the centers of the in-plane grid"""
+        return [(bs[1:] + bs[:-1]) / 2 for bs in self._plane_grid_boundaries]
```
[the remark this answers is not in the document]
Comment (lucas-flexcompute) [reply tidy3d/components/data/monitor_data.py:254]: These "2D" methods all depend on `_tangetial_dims`, which does the validation and raises a proper error message. We could implement the validation on each one, but it seems excessive, specially because those are private methods, so we should catch any errors during development.

File: tidy3d/components/data/monitor_data.py [hunk 19]
```diff
@@ -325,6 +375,22 @@ def flux(self) -> FluxDataArray:
         d_area = self._diff_area
         return FluxDataArray((self.poynting * d_area).sum(dim=d_area.dims))
 
+    @cached_property
+    def mode_area(self) -> FreqModeDataArray:
+        """Effective mode area corresponding to a 2D monitor
+
+        Effective mode area is calculated as: (∫|E|²dA)² / (∫|E|⁴dA)
```
[the remark this answers is not in the document]
Comment (lucas-flexcompute) [reply tidy3d/components/data/monitor_data.py:382]: Sorry, that's my fondness of Unicode showing… :)
I thought about using LaTeX for rendering in Sphinx but it would look bad in the terminal/notebook. Whatever we decide, I'm fine with it.

File: tidy3d/components/data/monitor_data.py [hunk 20]
```diff
@@ -325,6 +375,22 @@ def flux(self) -> FluxDataArray:
...
+        """
+        fields = self._centered_fields
+        abs_e_sq = fields["Ex"].abs ** 2 + fields["Ey"].abs ** 2 + fields["Ez"].abs ** 2
```
[the remark this answers is not in the document]
Comment (lucas-flexcompute) [reply tidy3d/components/data/monitor_data.py:385]: `_centered_fields` returns a dictionary, though. You mean something outside of the class, like so:

```python
def intensity(*components):
    return sum(np.abs(c) ** 2 for c in components)

class ElectromagneticFieldData(...)
    ...
    def mode_area(self):
        ...
        abs_e_qs = intensity(fields["Ex"], fields["Ey"], fields["Ez"])
        ....
```

I haven't found other places we could use that in this file, only in 2 other places in `sim_data.py`, so `intensity` could be defined somewhere else?

TBH I'm not really sure I like this change... sum of squares seems very straightforward to me I guess. And in `sim_data`, it's not always the `abs` that's used, so we can't use it directly everywhere.

File: tidy3d/components/data/monitor_data.py [hunk 21]
```diff
@@ -325,6 +375,22 @@ def flux(self) -> FluxDataArray:
...
+
+        # integrate over the plane
+        d_area = self._diff_area
+        num = (abs_e_sq * d_area).sum(dim=d_area.dims) ** 2
+        den = (abs_e_sq**2 * d_area).sum(dim=d_area.dims)
+
+        return FreqModeDataArray(num / den)
```
[the remark this answers is not in the document]
Comment (lucas-flexcompute) [reply tidy3d/components/data/monitor_data.py:392]: They are inherited directly from `num` and `den`, because they have the same coordinates.

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 22]
```diff
@@ -0,0 +1,944 @@
...
+        description="Solver grid resolution per wavelength.",
+    )
+
+    max_grid_scaling: float = pydantic.Field(
+        1.2,
+        title="Maximal Grid Scaling",
+        description="Maximal size increase between adjacent grid boundaries.",
+    )
+
+    compute_group_index: Union[bool, pydantic.PositiveFloat] = pydantic.Field(
```
[the remark this answers is not in the document]
Comment (lucas-flexcompute) [reply tidy3d/plugins/waveguide/rectangular_dielectric.py:193]: From a validation standpoint, the important fields are indeed in `ModeSpec`: `track_freq` and `precision`, so it might make sense.

File: tidy3d/plugins/waveguide/rectangular_dielectric.py [hunk 23]
```diff
@@ -0,0 +1,944 @@
...
+
+    Supports:
+    - Strip and rib geometries
+    - Angled sidewalls
+    - Modes in waveguide bends
+    - Surface and sidewall loss models
+    - Coupled waveguides
+    """
+
+    wavelength: Union[float, ArrayFloat1D] = pydantic.Field(
```
[the remark this answers is not in the document]
Comment (lucas-flexcompute) [reply tidy3d/plugins/waveguide/rectangular_dielectric.py:35]: Yes, the validator does that the conversion to a 1-element numpy array.

File: tidy3d/plugins/waveguide/rectangular_dielectric.py (hunk 13 above)
Comment (lucas-flexcompute) [reply tidy3d/plugins/waveguide/rectangular_dielectric.py:273]: I know the name [from OpenGL](https://www.khronos.org/opengl/wiki/Data_Type_(GLSL)#Swizzling), it's not my creation. That's a known operation to swap the axis order or add/remove/duplicate components. So, for the sake of clarity, I renamed to `_swap_axis`

**Comment by lucas-flexcompute** [comment]
I broke the long function into a few smaller ones. In the end it does look better indeed! Let me know what you think.

File: tidy3d/components/data/monitor_data.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (tylerflex) [reply tidy3d/components/data/monitor_data.py:375]: What I meant was something like 

```py
class ElectromagneticFieldData(...)
    ...
    @property
    def intensity(self):
        fields = self._centered_fields
        if any(cmp not in fields for cmp in ("Ex", "Ey", "Ez")):
            raise Exception("Can't compute intensity, all E fields must be present.")
        return = fields["Ex"].abs ** 2 + fields["Ey"].abs ** 2 + fields["Ez"].abs ** 2

    def mode_area(self):
        ...
        abs_e_qs = self.intensity
        ....
```

just adding this property, which could potentially be useful elsewhere. feel free to ignore if you disagree though

File: tidy3d/components/data/monitor_data.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (tylerflex) [reply tidy3d/components/data/monitor_data.py:382]: huh, didn't realize you could construct a `FreqModeDataArray` like that but good to know.

File: tidy3d/plugins/waveguide/rectangular_dielectric.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (tylerflex) [reply tidy3d/plugins/waveguide/rectangular_dielectric.py:195]: yea, sorry to refactor this again, but I do think this makes the most logical sense, especially given our previous discussion. Can we make this change here and in the group index calculation before that gets released? Thanks!

**Comment by tylerflex** [comment]
The refactored method looks much better, thanks!

I think the only additional comment (beyond the couple oustanding things, for example putting group index in ModeSpec) is just to give a bit of thought if possible to whether any of the structure creation can be generalized for the addition of other waveguide types (eg optical fibers). If we do make an abstract base class, I think it will be nice to have a few abstract methods to define what each new waveguide type will need to define (eg the override structures, grid spec, etc). But this is also probably something we can figure out when the time comes. 

Anyway, pending all of these things getting cleared up, I'll approve the PR.

**Review by tylerflex — APPROVED** [review]

**Comment by lucas-flexcompute** [comment]
I believe all changes are there. I've tried to comb all docstrings for lack of punctuation and I believe I've got all (most?) of them.
These changes are already in line with the `group_index_step` in `ModeSpec`.

**Comment by momchil-flex** [comment]
> These changes are already in line with the `group_index_step` in `ModeSpec`.

Sorry, what do you mean by that? When is this change going to come in?

Note that pylint is failing.

**Comment by lucas-flexcompute** [comment]
My bad, I forgot to add the latest commit before pinging you.

>> These changes are already in line with the group_index_step in ModeSpec.

> Sorry, what do you mean by that? When is this change going to come in?

They've already been merged to pre/2.1: https://github.com/flexcompute/tidy3d/pull/838

**Review by momchil-flex — APPROVED** [review]
